### PR TITLE
Enhanced Ubuntu-20 Dockerfile readability

### PR DESCRIPTION
**Pull request by Dunky-Z**
1. Delete some comment symbols such as ##;
2. Separate setting locale, update-alternatives and installing package;
3. One line per package;
4. Add a clean cache step; 
a. After installing the package, run apt clean, which reduces the size of the image package by about 113M.
5. Remove some invalid statements.Because each RUN instruction will add a layer, so if apt install and apt autoclean are separated, the cache generated by apt-get cannot be cleaned; 
a. remove RUN apt-get -y autoremove 
b. remove RUN apt-get clean

Signed-off-by: Dongdong Zhang <zhangdongdong@eswincomputing.com>

# Description

Please include a summary of the change and which issue is fixed or feature is
added.

Issue: #40

### Containers Affected

Please list the container images affected by this change. Including any
containers based on the targeted dockerfile.

### CODE REVIEW

**Comment by osteffenrh** [comment]
@Dunky-Z, thanks for taking an interest!

> 6. Remove the apt upgrade instruction, which downloads a lot of unnecessary content.

:thinking:  don't we want up-to-date packages? And if the base image is up-to-date it should not download anything new.

I am reluctant merging style-only changes since the value gain is strongly observer dependent.

Do we need a style guide for our dockerfiles?

@cfernald, @jgarver ?

**Comment by Dunky-Z** [comment]
@osteffenrh Thanks for replying. :heart:

I think it is only necessary to install the latest version of the software packages necessary to build the image, instead of updating all the software to the latest version. If the software is updated, the size of the image will increase. I think the upgrade operation should be handed over to the users who use the image, and let them upgrade on demand.

This PR is not only a style change, but also deletes some invalid commands.

The modification of the style is also consistent with the dockerfile of Fedora-35, so as to ensure that the style of the dockerfile of the entire project is consistent.

File: Ubuntu-20/Dockerfile [hunk 1]
```diff
@@ -25,42 +26,46 @@ ENV TZ=UTC
...
-
-# Install toolchains.
-RUN apt install -y \
-    g++-${GCC_MAJOR_VERSION} gcc-${GCC_MAJOR_VERSION} \
-    g++-${GCC_MAJOR_VERSION}-x86-64-linux-gnux32 gcc-${GCC_MAJOR_VERSION}-x86-64-linux-gnux32 \
-    g++-${GCC_MAJOR_VERSION}-aarch64-linux-gnu gcc-${GCC_MAJOR_VERSION}-aarch64-linux-gnu \
-    g++-${GCC_MAJOR_VERSION}-riscv64-linux-gnu gcc-${GCC_MAJOR_VERSION}-riscv64-linux-gnu \
-    g++-${GCC_MAJOR_VERSION}-arm-linux-gnueabi gcc-${GCC_MAJOR_VERSION}-arm-linux-gnueabi \
-    g++-${GCC_MAJOR_VERSION}-arm-linux-gnueabihf gcc-${GCC_MAJOR_VERSION}-arm-linux-gnueabihf && \
+    echo "deb https://download.mono-project.com/repo/ubuntu stable-focal main" | sudo tee /etc/apt/sources.list.d/mono-official-stable.list &&\
```
Comment (jgarver) [inline Ubuntu-20/Dockerfile:55]: Not sure I like hiding the mono PPA in here.

The steps to install mono were a direct lift from stuart docs.  I think it's best to keep them that way so it's easy to update them as the upstream docs change.

File: Ubuntu-20/Dockerfile [hunk 2]
```diff
@@ -97,36 +102,41 @@ RUN apt install -y \
...
 
 # Install convenience tools.  Things we like having around, but aren't
 # required.
-RUN apt-get install -y vim nano bear
+RUN apt-get install --yes --no-install-recommends \
+        vim \
+        nano \
+        bear &&\
+    apt-get clean &&\
+    rm -rf /var/lib/apt/lists/*
```
Comment (jgarver) [inline Ubuntu-20/Dockerfile:139]: Would be nice to include this cleanup at the end up the build image, too.  I'm more worried about build image size than dev image.  That would mean doing another apt update to start the dev image, though.

**Review by jgarver — COMMENTED** [review]
I don't mind the style updates and the changes to improve the image size.

However, I strongly disagree with removing "apt upgrade".  On the build image, that would force me to do an "apt upgrade" during each of my builds.  On the dev image, I think you're forcing users to FROM this image and build their own image instead of using this one directly.

File: Ubuntu-20/Dockerfile [hunk 3]
```diff
@@ -25,42 +26,46 @@ ENV TZ=UTC
...
-# Set the locale.
-RUN apt-get install --yes --no-install-recommends locales && \
-    sed -i '/en_US.UTF-8/s/^# //g' /etc/locale.gen && \
-    locale-gen
-ENV LANG en_US.UTF-8
-ENV LANGUAGE en_US:en
-ENV LC_ALL en_US.UTF-8
-
-# Upgrade all packages.
-RUN apt-get upgrade -y
```
[the remark this answers is not in the document]
Comment (Dunky-Z) [reply Ubuntu-20/Dockerfile:40]: Thanks for your comment. 
Ok, I keep the steps of apt-get upgrade.

File: Ubuntu-20/Dockerfile (hunk 1 above)
Comment (Dunky-Z) [reply Ubuntu-20/Dockerfile:55]: Ok, I'll keep them.

File: Ubuntu-20/Dockerfile (hunk 2 above)
Comment (Dunky-Z) [reply Ubuntu-20/Dockerfile:139]: Yep, I tried to cleanup in the build image before, but in the dev image I needed to re-update. so I decided to cleanup in the dev image, we just make sure that the installed package can be developed properly, if the developer has other needs, do the apt update operation by himself, it's not a troublesome operation.

**Comment by osteffenrh** [comment]
With package updates I'll approve.

@Dunky-Z, build is failing, can you check?

**Comment by Dunky-Z** [comment]
> With package updates I'll approve.
> 
> @Dunky-Z, build is failing, can you check?

Thanks, I'll check it.

**Comment by Dunky-Z** [comment]
I have solved it, could you please approve the workflow(Action)? @osteffenrh :grin:

**Comment by cfernald** [comment]
I'm okay with style only changes, though it seems this is more than just style anyways. The added cleans during the apt-get install RUN directives should also help reduce the image and layer sizes too.

**Comment by Dunky-Z** [comment]
Hi @osteffenrh ,Could this patch be merged in?

**Pull request merged**